### PR TITLE
Fix storage tests on Python 3

### PR DESCRIPTION
**Pull request by claudep**
Only remaining storage test failures on Python 3 are due to #3322.

### CODE REVIEW

File: translate/storage/test_factory.py [hunk 1]
```diff
@@ -109,9 +109,8 @@ def test_gzfile(self):
 
     def test_bz2file(self):
         """Test that we can open a gzip file correctly."""
-        if not BZ2File:
-            return
         filename = os.path.join(self.testdir, self.filename + '.bz2')
+        # `with` support requires Python 2.7
```
Comment (unho) [inline translate/storage/test_factory.py]: Sorry?

Comment (claudep) [inline translate/storage/test_factory.py]: I meant `with open(...) as` with a BZ2File object (https://docs.python.org/2/library/bz2.html?highlight=bz2#bz2.BZ2File).

File: translate/storage/csvl10n.py [hunk 2]
```diff
@@ -416,12 +415,11 @@ def serialize(self):
...
     def getoutput(self):
-        outputfile = BytesIO()
-        writer = csv.DictWriter(outputfile, self.fieldnames, extrasaction='ignore', dialect=self.dialect)
-        # write header
-        hdict = dict(map(None, self.fieldnames, self.fieldnames))
-        writer.writerow(hdict)
+        output = csv.StringIO()
+        writer = csv_utils.UnicodeDictWriter(
+            output, self.fieldnames, encoding=self.encoding, extrasaction='ignore',
+            dialect=self.dialect)
```
Comment (unho) [inline translate/storage/csvl10n.py]: Better:

```python
        writer = csv_utils.UnicodeDictWriter(output, self.fieldnames,
                                             encoding=self.encoding,
                                             extrasaction='ignore',
                                             dialect=self.dialect)
```

Comment (claudep) [inline translate/storage/csvl10n.py]: Better is subjective, let's say it's your preferred syntax :-) Both are accepted by pep8.

File: translate/storage/mo.py [hunk 3]
```diff
@@ -264,16 +265,18 @@ def parse(self, input):
...
-                charset = re.search("charset=([^\\s]+)",
+                charset = re.search(b"charset=([^\\s]+)",
                                     input[voffset:voffset + vlength])
                 if charset:
                     self.encoding = charset.group(1)
-            target = multistring(input[voffset:voffset + vlength].split("\0"))
+            source = multistring(
+                [s.decode(self.encoding) for s in source.split(b"\0")])
+            target = multistring(
+                [s.decode(self.encoding) for s in input[voffset:voffset + vlength].split(b"\0")])
```
Comment (unho) [inline translate/storage/mo.py]: I find preferable the following:

```python
            source = multistring([s.decode(self.encoding)
                                  for s in source.split(b"\0")])
            target = multistring([s.decode(self.encoding)
                                  for s in input[voffset:voffset + vlength].split(b"\0")])
```

File: translate/storage/csvl10n.py (hunk 2 above)
Comment (unho) [inline translate/storage/csvl10n.py]: Sure. But lets try to align with our styleguide: http://docs.translatehouse.org/projects/translate-toolkit/en/latest/developers/styleguide.html#continuing-long-statements

File: translate/storage/test_poheader.py [hunk 4]
```diff
@@ -51,10 +51,9 @@ def test_update():
     d['Project-Id-Version'] = 'abc'
     d['POT-Creation-Date'] = 'now'
     d = poheader.update(d, add=True, Test='hello', Report_Msgid_Bugs_To='bugs@list.org')
-    assert d.keys()[0] == "Project-Id-Version"
-    assert d.keys()[1] == "Report-Msgid-Bugs-To"
-    assert d.keys()[2] == "POT-Creation-Date"
-    assert d.keys()[3] == "Test"
+    assert list(d.keys()) == [
+        "Project-Id-Version", "Report-Msgid-Bugs-To", "POT-Creation-Date", "Test",
```
Comment (unho) [inline translate/storage/test_poheader.py]: The last element is beyond the preferred maximum line length, so please wrap it.

File: translate/storage/test_po.py [hunk 5]
```diff
@@ -463,26 +476,26 @@ def test_obsolete(self):
...
+        posource = u'''msgid ""
 msgstr ""
-"PO-Revision-Date: 2006-02-09 23:33+0200\n"
-"MIME-Version: 1.0\n"
-"Content-Type: text/plain; charset=UTF-8\n"
-"Content-Transfer-Encoding: 8-bit\n"
+"PO-Revision-Date: 2006-02-09 23:33+0200\\n"
+"MIME-Version: 1.0\\n"
+"Content-Type: text/plain; charset=UTF-8\\n"
+"Content-Transfer-Encoding: 8-bit\\n"
```
Comment (unho) [inline translate/storage/test_po.py]: Another of these changes I don't like. Is it possible to avoid it?

File: translate/storage/cpo.py [hunk 6]
```diff
@@ -525,19 +531,19 @@ def setmsgidcomment(self, msgidcomment):
...
         location = gpo.po_message_filepos(self._gpo_message, i)
         while location:
-            locname = gpo.po_filepos_file(location).decode(self.CPO_ENC)
+            locname = gpo_decode(gpo.po_filepos_file(location))
             locline = gpo.po_filepos_start_line(location)
             if locline == -1:
                 locstring = locname
             else:
-                locstring = locname + u":" + six.text_type(locline)
+                locstring = u"%s:%s" % (locname, six.text_type(locline))
```
Comment (unho) [inline translate/storage/cpo.py]: Perhaps more readable:

```python
                locstring = u":".join(locname, six.text_type(locline))
```

**Comment by unho** [comment]
lgtm after comments addressed.

File: translate/storage/csvl10n.py (hunk 2 above)
Comment (claudep) [inline translate/storage/csvl10n.py]: I'll fix that, but I maintain my preference for the other syntax, which is more adequate when we want to keep shorter lines. Styleguides can change also :-)

**Pull request merged**